### PR TITLE
output data source samples

### PR DESCRIPTION
**Pull request by qiqingjin**

### CODE REVIEW

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+4. In `widget.tsx`, use `DataSourceComponent` to create original data source instance.
+5. Update output data source's source records every time user clicking `update output data source` button.
+
+```tsx
+<Button onClick={this.setSourceRecordsToOutputDs}>
+  Update output data source
+</Button>
+```
+
+Please note that the widget only generates output data source. If you want to use the output data source, please add another widget (such as List widget) and select the output data source for it.
```
Comment (qlqllu) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:46]:
```suggestion
Please note that the widget only generates output data source. If you want to use the output data source, please add another widget (such as List widget) and select the output data source for it. This sample widget does not listen to the origin data source change, you can use `onDataSourceInfoChange` callback of the `DataSourceComponent` to listen to the origin data source change.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+4. In `widget.tsx`, use `DataSourceComponent` to create original data source instance.
+5. Update output data source's source records every time user clicking `update output data source` button.
+
+```tsx
+<Button onClick={this.setSourceRecordsToOutputDs}>
+  Update output data source
+</Button>
+```
+
```
Comment (qlqllu) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md:54]:
```suggestion
6. The main difference between this widget with the `client-side-output` is this widget's output data source uses a different schema with its origin data source.
```

**Comment by qlqllu** [comment]
This should be to `1.4` branch.

File: samples/widgets/widget-output-data-source/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+A widget can use data source and it can generate data source as well, we call the generated data source as widget output data source. Widgets can use output data sources the way same as using other data sources. 
+
+## How to generate output data source
+
+In general, there are two steps for widget to generate output data source.
+
+1. In setting, declare its output data source by calling `this.props.onSettingChange`.
+2. In widget, update data of output data source.
+
+More specifically, output data source can save data in data source instance, or not. If save data in data source instance, widget needs to update output data source's source records, if not, widget needs to update output data source's query parameter. To finish step 2, there are two options.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/README.md:12]:
```suggestion
More specifically, an output data source can save data in a data source instance, or not. If your widget needs to save data in a data source instance, use `outputDataSource.setSourceRecords`. If your widget does not need to save data in the data source instance, use `outputDataSource.updateQueryParams`.
```

File: samples/widgets/widget-output-data-source/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+2. In widget, update data of output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/README.md:14]:
```suggestion

```

File: samples/widgets/widget-output-data-source/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+   1. If save data in data source instance, call `outputDataSource.setSourceRecords`.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/README.md:15]:
```suggestion

```

File: samples/widgets/widget-output-data-source/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+   2. If don't save data in data source instance, call `outputDataSource.updateQueryParams`.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/README.md:16]:
```suggestion

```

File: samples/widgets/widget-output-data-source/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+## How to use output data source
+
+Widgets can use output data source just like using any other data sources.
+
+1. In setting, use `DataSourceSelector` to let user select a data source (`Outputs` tab).
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/README.md:22]:
```suggestion
1. In the setting, use `DataSourceSelector` to let user select a data source (`Outputs` tab).
```

File: samples/widgets/widget-output-data-source/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+2. In widget, use `DataSourceComponent` to create data source instance and use the data source instance to do other operations.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/README.md:23]:
```suggestion
2. In the widget, use `DataSourceComponent` to create data source instance and use the data source instance to do other operations.
```

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
+# Client-side output data source
+
+Output data source can save data in data source instance, we call it client-side output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:3]:
```suggestion
An output data source can save data in data source instance, this is called a client-side output data source.
```

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+## How to use the sample
+
+Clone the [sample repo](https://github.com/esri/arcgis-experience-builder-sdk-resources) and copy this widget's folder (within `samples/widgets`) to the `client/your-extensions/widgets` folder of your Experience Builder installation.
+
+## How it works
+
+1. In `setting.tsx`, use `DataSourceSelector` to allow user to select an original data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:11]:
```suggestion
1. In `setting.tsx`, use `DataSourceSelector` to allow the user to select an original data source.
```

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+2. In `setting.tsx`, declare output data source inside `onChange` callback of `DataSourceSelector`.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:13]:
```suggestion
2. In `setting.tsx`, declare output data source inside `onChange` callback of `DataSourceSelector`.
```

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+```ts
+// Let framework know which data source current widget is using and which data source current widget is outputing.
+this.props.onSettingChange({
+  id: this.props.id,
+  useDataSources: useDataSources
+}, outputDsJsons)
+```
+
+3. In `setting.tsx`, select a filter and save it in widget config.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:23]:
```suggestion
3. In `setting.tsx`, select a filter and save it in the widget config.
```

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  mode={SqlExpressionMode.Simple}
+  expression={this.props.config.sqlExpression}
+  // use this.props.onSettingChange to save filter to config inside this callback
+  onChange={this.onSqlExpressionChange}
+  isOpen={this.state.isSqlBuilderOpen}
+  toggle={this.toggleSqlBuilder}
+/>
+```
+
+4. In `widget.tsx`, use `DataSourceComponent` to create original data source instance.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:37]:
```suggestion
4. In `widget.tsx`, use `DataSourceComponent` to create the original data source instance.
```

File: samples/widgets/widget-output-data-source/client-side-output/README.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+5. Update output data source's source records every time user clicking `update output data source` button.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/README.md:38]:
```suggestion
5. Update output data source's source records every time the user clicks the  `update output data source` button.
```

File: samples/widgets/widget-output-data-source/server-side-output/README.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
+# Server-side output data source
+
+Output data source may not save data in data source instance, but uses remote database, we call it server-side output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/README.md:3]:
```suggestion
An output data source may not save data in a data source instance but uses a remote database,  this is called a server-side output data source.
```

File: samples/widgets/widget-output-data-source/server-side-output/README.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+## How to use the sample
+
+Clone the [sample repo](https://github.com/esri/arcgis-experience-builder-sdk-resources) and copy this widget's folder (within `samples/widgets`) to the `client/your-extensions/widgets` folder of your Experience Builder installation.
+
+## How it works
+
+1. In `setting.tsx`, use `DataSourceSelector` to allow user to select an original data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/README.md:11]:
```suggestion
1. In `setting.tsx`, use `DataSourceSelector` to allow the user to select an original data source.
```

File: samples/widgets/widget-output-data-source/server-side-output/README.md [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+```ts
+// Let framework know which data source current widget is using and which data source current widget is outputing.
+this.props.onSettingChange({
+  id: this.props.id,
+  useDataSources: useDataSources
+}, outputDsJsons)
+```
+
+3. In `setting.tsx`, select a filter and save it in widget config.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/README.md:23]:
```suggestion
3. In `setting.tsx`, select a filter and save it in the widget config.
```

File: samples/widgets/widget-output-data-source/server-side-output/README.md [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  expression={this.props.config.sqlExpression}
+  // use this.props.onSettingChange to save filter to config inside this callback
+  onChange={this.onSqlExpressionChange}
+  isOpen={this.state.isSqlBuilderOpen}
+  toggle={this.toggleSqlBuilder}
+/>
+```
+
+4. In `widget.tsx`, use `DataSourceComponent` to create original data source instance.
+5. Update output data source's query parameters every time user clicking `update output data source` button.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/README.md:38]:
```suggestion
5. Update the output data source's query parameters every time the user clicks the `update output data source` button.
```

File: samples/widgets/widget-output-data-source/server-side-output/README.md [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+```tsx
+<Button onClick={this.updateOutputDsQueryParams}>
+  Update output data source
+</Button>
+```
+
+Please note that the widget only generates output data source. If you want to use the output data source, please add another widget (such as List widget) and select the output data source for it.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/README.md:46]:
```suggestion
Please note that the widget only generates an output data source. If you want to use the output data source, please add another widget (such as the List widget) and select the output data source for it.
```

File: samples/widgets/widget-output-data-source/server-side-output/src/runtime/widget.tsx [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    featureLayerOutputDs.setStatus(DataSourceStatus.Unloaded)
+    featureLayerOutputDs.setCountStatus(DataSourceStatus.Unloaded)
+  }
+
+  render () {
+    if (!this.isDsConfigured()) {
+      return (
+        <h3>
+          This widget demostrates how to use attribute query result to generate server-side output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/src/runtime/widget.tsx:66]:
```suggestion
          This widget demonstrates how to use attribute query results to generate a server-side output data source.
```

File: samples/widgets/widget-output-data-source/client-side-output/src/runtime/widget.tsx [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+      this.getOutputDataSource()?.setStatus(DataSourceStatus.Unloaded)
+      this.getOutputDataSource()?.setCountStatus(DataSourceStatus.Unloaded)
+    })
+  }
+
+  render () {
+    if (!this.isDsConfigured()) {
+      return (
+        <h3>
+          This widget demostrates how to use attribute query result to generate client-side output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/src/runtime/widget.tsx:72]:
```suggestion
          This widget demonstrates how to use attribute query results to generate a client-side output data source.
```

File: samples/widgets/widget-output-data-source/client-side-output/src/runtime/widget.tsx [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+    return (
+      <div className='widget-attribute-query-result-as-output' style={{ width: '100%', height: '100%', maxHeight: '800px', overflow: 'auto' }}>
+        <h3>
+          The client-side output data source is generated.
+        </h3>
+        <h5>
+          If you want to use the client-side output data source, you should add another widget (such as List) and configure the output data source for it.
+        </h5>
+        <h5>
+          If the widget's origin data source is changed (e.g. add a filter), output data source won't update automotically. Please click the following button.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/client-side-output/src/runtime/widget.tsx:88]:
```suggestion
          If the widget's origin data source is changed (e.g. add a filter), the output data source won't update automatically. Please click the following button.
```

File: samples/widgets/widget-output-data-source/server-side-output/src/runtime/widget.tsx [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+    return (
+      <div className='widget-attribute-query-result-as-output' style={{ width: '100%', height: '100%', maxHeight: '800px', overflow: 'auto' }}>
+        <h3>
+          The server-side output data source is generated.
+        </h3>
+        <h5>
+          If you want to use the server-side output data source, you should add another widget (such as List) and configure the output data source for it.
+        </h5>
+        <h5>
+          If the widget's origin data source is changed (e.g. add a filter), output data source won't update automotically. Please click the following button.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/server-side-output/src/runtime/widget.tsx:82]:
```suggestion
          If the widget's origin data source is changed (e.g. add a filter), the output data source won't update automatically. Please click the following button.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md [hunk 24]
```diff
@@ -0,0 +1,56 @@
+# Client-side output data source
+
+Output data source can save data in data source instance, we call it client-side output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md:3]:
```suggestion
An output data source can save data in a data source instance, this is called a client-side output data source.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+## How to use the sample
+
+Clone the [sample repo](https://github.com/esri/arcgis-experience-builder-sdk-resources) and copy this widget's folder (within `samples/widgets`) to the `client/your-extensions/widgets` folder of your Experience Builder installation.
+
+## How it works
+
+1. In `setting.tsx`, use `DataSourceSelector` to allow user to select an original data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md:11]:
```suggestion
1. In `setting.tsx`, use `DataSourceSelector` to allow the user to select an original data source.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+<MultiSelect fluid
+  items={this.getStatFuncItems()}
+  values={this.getSelectedStatFuncs()}
+  // use this.props.onSettingChange to save statistic funtions to config inside this callback
+  onClickItem={this.onStatFuncItemClick}
+/>
+```
+
+4. In `widget.tsx`, use `DataSourceComponent` to create original data source instance.
+5. Update output data source's source records every time user clicking `update output data source` button.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/README.md:47]:
```suggestion
5. Update output data source's source records every time the user clicks the `update output data source` button.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/src/runtime/widget.tsx [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+      console.error(`${this.props.id} loading data error, `, err)
+    })
+  }
+
+  render () {
+    if (!this.isDsConfigured()) {
+      return (
+        <>
+          <h3>
+            This widget demostrates how to use statistic query result to generate client-side output data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/src/runtime/widget.tsx:133]:
```suggestion
            This widget demonstrates how to use statistic query results to generate a client-side output data source.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/src/runtime/widget.tsx [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+          </h3>
+          <h4>
+            Firstly, please config data source.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/src/runtime/widget.tsx:136]:
```suggestion
            Please config data source.
```

File: samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/src/runtime/widget.tsx [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+    return (
+      <div className='widget-use-feature-set' style={{ width: '100%', height: '100%', maxHeight: '800px', overflow: 'auto' }}>
+        <h3>
+          The output data source is generated.
+        </h3>
+        <h5>
+          The widget will use its output data source and will list the records in the output data source.
+        </h5>
+        <h5>
+          If the widget's origin data source is changed (e.g. add a filter), output data source won't update automotically. Please click the following button.
```
Comment (dmartine) [inline samples/widgets/widget-output-data-source/statistic-result-as-client-side-output/src/runtime/widget.tsx:153]:
```suggestion
          If the widget's origin data source is changed (e.g. add a filter), the output data source won't update automatically. Please click the following button.
```

**Comment by qiqingjin** [comment]
Committed all above suggestions and moved to the PR to 1.4 branch. https://github.com/Esri/arcgis-experience-builder-sdk-resources/pull/62 .

**Pull request closed without merging**